File: src/thrower-identity.test.js
import test from 'node:test';
import assert from 'node:assert/strict';
import {initial,catchBall,step} from './game.js';

test('catch preserves the last thrower until the catcher begins a new windup',()=>{
 for(const team of [0,1]) {
  const s=initial(), catcher=team*9+7, thrower=(1-team)*9+7;
  for(const p of s.players) Object.assign(p,{x:20,z:10,stun:100});
  Object.assign(s.players[catcher],{x:0,z:0,stun:0});
  s.controlled[team]=catcher;
  Object.assign(s.ball,{x:.1,z:0,h:.75,vx:1,owner:-1,lastTouch:thrower});
  catchBall(s,catcher);
  assert.equal(s.ball.owner,catcher);
  assert.equal(s.ball.lastTouch,thrower);
  for(let n=0;n<5 && s.players[catcher].action!==3;n++)
   step(s,.04,team===0?{shoot:true}:{},[true,true],team===1?{shoot:true}:{});
  assert.equal(s.players[catcher].action,3);
  assert.equal(s.ball.owner,catcher,'identity changes before release');
  assert.equal(s.ball.lastTouch,catcher);
 }
});

File: server/thrower_identity_test.go
package main

import "testing"

func TestThrowerIdentityAtWindup(t *testing.T) {
	for team := 0; team < 2; team++ {
		s := initial()
		catcher, thrower := team*9+7, (1-team)*9+7
		for i := range s.Players {
			s.Players[i].X, s.Players[i].Z, s.Players[i].Stun = 20, 10, 100
		}
		p := &s.Players[catcher]
		p.X, p.Z, p.Stun = 0, 0, 0
		s.Controlled[team] = catcher
		s.Ball = Ball{X: .1, H: .75, VX: 1, Owner: -1, LastTouch: thrower}
		s.catchBall()
		if s.Ball.Owner != catcher || s.Ball.LastTouch != thrower {
			t.Fatal("catch changed thrower", s.Ball)
		}
		for n := 0; n < 5 && p.Action != 3; n++ {
			inputs := [2]Input{}
			inputs[team].Shoot = true
			s.simulate(.04, inputs, [2]bool{true, true})
		}
		if p.Action != 3 || s.Ball.Owner != catcher || s.Ball.LastTouch != catcher {
			t.Fatal("windup identity", s.Ball, p.Action)
		}
	}
}
